HoistFail

lrworth opened the pull request:
Adds a new typeclass `HoistFail` analogous to `HoistError` but for hoisting into `MonadFail`. This is concretely useful for writing aeson and cassava parsers; both of these libraries use `MonadFail` to support parse errors, and our parsers tend to be littered with variations of `either fail pure`.

To support common use cases I've added some functions with mnemonic suffixes:
- a family of functions ending with `'`, which simplify hoisting from different monads with the same error type
- a family of functions with `E` in the suffix, which set the input monad to `Either` to support hoisting from types where the monad is polymorphic.

lrworth committed:
feat: Control.Monad.Fail.Hoist

lrworth committed:
feat: hoistFailString and hoistFailShow

lrworth committed:
feat! hoistFailString -> hoistFail_

lrworth committed:
feat: reorder type variables to make t settable with type var

lrworth committed:
feat: add hoistFailE and hoistFailE_

Useful to fix the type of the monad so that it works with polymorphic
functions.

lrworth commented on src/Control/Monad/Fail/Hoist.hs:
```diff
@@ -0,0 +1,83 @@
+hoistFail_ :: HoistFail t m String => t a -> m a
+hoistFail_ = hoistFail id
+
+hoistFailE :: MonadFail m => (e -> String) -> Either e a -> m a
+hoistFailE = hoistFail
```
This makes it easier to fix the monad on polymorphic functions, otherwise it is simpler to use `either`. I wouldn't mind a similar `hoistErrorE`.

endgame commented:
I would be OK with dropping support for ancient GHC.

lrworth committed:
chore: drop support for GHC versions prior to MonadFail

lrworth committed:
fix: old GHC did not include MonadFail in Prelude

JackKelly-Bellroy commented on src/Control/Monad/Fail/Hoist.hs:
```diff
@@ -0,0 +1,85 @@
+instance MonadFail m => HoistFail (Either e) m e where
+  hoistFail f = either (fail . f) pure
+
+hoistFail_ :: HoistFail t m String => t a -> m a
+hoistFail_ = hoistFail id
```
The `_` suffix usually means "discard result" to me; I'd probably call this `hoistFail'`.

JackKelly-Bellroy replied:
And maybe we want `hoistFailM`/`hoistFailM'`/`hoistFailEM`/`hoistFailEM'` for consistency with `hoistError` functions?

lrworth replied:
I've added many functions.

JackKelly-Bellroy commented on src/Control/Monad/Fail/Hoist.hs:
```diff
@@ -0,0 +1,85 @@
+-- ('<%!>') :: 'MonadFail' m => 'Maybe'       a -> (() -> e) ->           m a
+-- ('<%!>') :: 'MonadFail' m => 'Either'  a   b -> (a  -> e) ->           m b
+-- @
+(<%!>)
```
Bikeshed: I am not thrilled by these operator names, but I cannot think of better ones. `(<!>)` is the semigroupoids version of `(<|>)`, and the `!` is already used in `Control.Monad.Error.Hoist` to mean "with monadic effect".

`#` kinda looks like a bad 'F', and there [aren't very many occurrences of `(<#>)`](https://hoogle.haskell.org/?hoogle=%3C%23%3E) to mess with an existing mnemonic. WDYT about `(<%#>)`, `(<%!#>)`, `(<#>)`, and `<#!>`)?

lrworth replied:
Sure, I'll change it.

lrworth committed:
Bikeshed function and operator names

lrworth committed:
feat: add functions for parity with Error.Hoist

lrworth committed:
doc: add haddock to Fail.Hoist

lrworth committed:
doc: fix arrow in Error.Hoist

lrworth committed:
feat: add functions to Error.Hoist for parity with Fail.Hoist

lrworth committed:
Order HoistFail type params same as HoistError

`hoistFailE` and co mean you don't have to use type applications as
often, so order doesn't matter as much. The alternative to this was
changing the order of `HoistError` params for consistency, but this
would not be backward-compatible and would require a major version bump.

lrworth commented on src/Control/Monad/Error/Hoist.hs:
```diff
@@ -70,9 +76,9 @@ class Monad m => HoistError m t e e' | t -> e where
-  -- 'hoistError' :: 'MonadError' e m -> (() -> e) -> 'Maybe'       a -> m a
-  -- 'hoistError' :: 'MonadError' e m -> (a  -> e) -> 'Either'  a   b -> m b
-  -- 'hoistError' :: 'MonadError' e m -> (a  -> e) -> 'ExceptT' a m b -> m b
+  -- 'hoistError' :: 'MonadError' e m => (() -> e) -> 'Maybe'       a -> m a
+  -- 'hoistError' :: 'MonadError' e m => (a  -> e) -> 'Either'  a   b -> m b
+  -- 'hoistError' :: 'MonadError' e m => (a  -> e) -> 'ExceptT' a m b -> m b
```
This has been a long time coming.

endgame commented on src/Control/Monad/Error/Hoist.hs:
```diff
@@ -120,6 +157,36 @@ hoistErrorM e m = do
+  -> m a
+hoistErrorM' m = do
+  x <- m
+  hoistError' x
```
```suggestion
hoistErrorM' m = m >>= hoistError'
```

endgame reviewed (commented):
I keep wondering, "what if the new typeclasses aren't necessary?" From there, I wonder if the basic design of this library could be better?

The core class of `hoist-error` is:
```haskell
class Monad m => HoistError m t e e' | t -> e where
  hoistError :: (e -> e') -> t a -> m a
```

It tells us which error/partiality functor `t` is in play (e.g., `Maybe`, `Either e`, `ExceptT e m`), what the error type `e` is for `t` (`Maybe` -> `()`; `Either e` -> `e`) and which error type is used by the monad `m`. If we separate the ideas of "identifying the error" and "hoisting the error into some other monad", then maybe we can avoid rewriting all machinery for `MonadError` and `MonadFail`. Perhaps something like this?

```haskell
class InspectError e t m | t -> e where
  inspectError :: t a -> m (Either e a)

instance Applicative m => InspectError () Maybe m where
  inspectError = pure . \case
    Nothing -> Left ()
    Just a -> Right a

instance Applicative m => InspectError e (Either e) m where
  inspectError = pure

instance InspectError e (ExceptT e m) m where
  inspectError = runExceptT

hoistError :: (MonadError e' m, InspectError e t m) => (e -> e') -> t a -> m a
hoistError f = either (throwError . f) pure <=< inspectError

hoistFail :: (MonadFail m, InspectError e t m) => (e -> String) -> t a -> m a
hoistFail f = either (fail . f) pure <=< inspectError
```

I'd probably put the `InspectError` class in `Control.Monad.Error.Hoist`, put all the `fail`-based machinery in `Control.Monad.Fail.Hoist`, and then everything _should_ "just work".

lrworth committed:
refactor: remove do-notation

lrworth commented:
> I wonder if the basic design of this library could be better?

Sounds great, but it'll be a breaking change so should happen in a different PR.

JackKelly-Bellroy commented:
In the spirit of "make the hard change easy, then make the easy change", I think we should look at the `InspectError` refactor first.

lrworth committed:
fix: signatures of <#> and <#!> forcing String error context

In particular this meant the operators didn't work with Maybe.
